File: ios/Sources/CompanionCore/ThreadNavigation.swift
import Foundation

/// One folder's visible threads, or the unfiled threads after the folders.
public struct BotThreadGroup: Identifiable, Hashable, Sendable {
    public let project: BotProject?
    public let tasks: [BotTask]

    public var id: String { project.map { "project:\($0.id)" } ?? "unfiled" }
}

extension BotTask {
    public var displayTitle: String {
        let trimmed = title.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? "Untitled thread" : trimmed
    }
}

extension Bot {
    /// Saved folder order and server thread order are preserved. Missing
    /// folders leave their threads accessible in the unfiled group, and
    /// within every group attention outranks recency: a thread that needs
    /// the person floats above the idle tail, which keeps stored order.
    /// A folder-name search keeps all of that folder's visible threads,
    /// in relevance order rather than attention tiers.
    ///
    /// Threads a bot closed are folded away by default, the way the desktop
    /// sidebar folds them: a PM bot that opened ten helper threads and closed
    /// them must not leave ten rows behind. They are never gone — a search
    /// or `includingClosed` (the manage sheet) lists them, and a closed
    /// thread that is running, unread, or open here stays in the list. A
    /// thread the person archived folds away the same way, with the same
    /// attention override.
    /// - Parameter queuedThreadIds: threads holding a queued send, from the
    ///   client's queue state. A closed or archived thread with a held send
    ///   stays in the list the way a running one does — activity strings
    ///   never say this, because the harness reports queues out-of-band.
    public func threadGroups(
        matching query: String = "",
        includingClosed: Bool = false,
        queuedThreadIds: Set<String> = []
    ) -> [BotThreadGroup] {
        let search = query.trimmingCharacters(in: .whitespacesAndNewlines)
        let threads: [BotTask]
        if tasks == nil {
            // Older computers have one conversation but no task metadata.
            // An explicitly empty modern list must stay empty.
            threads = [BotTask(
                threadId: threadId, title: "", createdAt: createdAt,
                modelSelection: modelSelection, busy: busy, waitingOnTeammate: waitingOnTeammate,
                unread: unread,
                approvalMode: approvalMode, autoApprove: autoApprove, alwaysAllow: alwaysAllow
            )]
        } else if includingClosed || !search.isEmpty {
            threads = visibleTasks
        } else {
            threads = visibleTasks.filter { task in
                !(task.isClosed || task.isArchived)
                    || task.demandsAttention(queued: queuedThreadIds.contains(task.threadId))
                    || task.threadId == threadId
            }
        }
        let ordered = search.isEmpty ? threadsInAttentionOrder(threads, queuedThreadIds: queuedThreadIds) : threads

        var projectIDs = Set<String>()
        var groups = (projects ?? []).compactMap { project -> BotThreadGroup? in
            guard projectIDs.insert(project.id).inserted else { return nil }
            let filed = ordered.filter { $0.projectId == project.id }
            return filed.isEmpty ? nil : BotThreadGroup(project: project, tasks: filed)
        }
        let unfiled = ordered.filter { task in
            task.projectId.map { !projectIDs.contains($0) } ?? true
        }
        if !unfiled.isEmpty {
            groups.append(BotThreadGroup(project: nil, tasks: unfiled))
        }

        guard !search.isEmpty else { return groups }
        return groups.compactMap { group in
            if group.project?.name.localizedStandardContains(search) == true { return group }
            let matches = group.tasks.filter { $0.displayTitle.localizedStandardContains(search) }
            return matches.isEmpty ? nil : BotThreadGroup(project: group.project, tasks: matches)
        }
    }

    /// Attention outranks recency within a bot: waiting-on-you needs the
    /// person most, then working/busy, then queued, then unread. A held send
    /// is client state, so it ranks in the queued tier the way the wire
    /// value does. The thread being looked at rides just above the idle
    /// tail; idle threads keep stored order. Mirrors the desktop's
    /// attentionRank so the tree and the manage sheet agree on what sits on
    /// top; searches keep relevance order, as on desktop and Android.
    private func attentionRank(_ task: BotTask, queued: Bool) -> Int {
        if task.activity == "waiting-on-you" { return 0 }
        if task.busy == true || task.activity == "working" { return 1 }
        if task.activity == "queued" || queued { return 2 }
        if task.unread == true { return 3 }
        if task.threadId == threadId { return 4 }
        return 5
    }

    /// Order, never filter: whatever the caller passed stays in the list,
    /// only its position changes. The stored index rides along so equal
    /// ranks keep stored order even where sort is not guaranteed stable.
    private func threadsInAttentionOrder(
        _ threads: [BotTask],
        queuedThreadIds: Set<String>
    ) -> [BotTask] {
        threads.enumerated()
            .map { (index: $0.offset, task: $0.element) }
            .sorted {
                let lhs = attentionRank($0.task, queued: queuedThreadIds.contains($0.task.threadId))
                let rhs = attentionRank($1.task, queued: queuedThreadIds.contains($1.task.threadId))
                return lhs == rhs ? $0.index < $1.index : lhs < rhs
            }
            .map(\.task)
    }
}

File: android/core/src/main/kotlin/com/openmausbot/companion/core/ThreadNavigation.kt
package com.openmausbot.companion.core

/** A saved folder's visible threads, or the unfiled threads after the folders. */
data class BotThreadGroup(val project: BotProject?, val tasks: List<BotTask>) {
    val id: String get() = project?.let { "project:${it.id}" } ?: "unfiled"
}

val BotTask.displayTitle: String
    get() = title.trim().ifEmpty { "Untitled thread" }

/** The thread's own turn is running — the desktop's isWorking exactly.
 * A run counts as work here exactly as its row labels it Working. */
val BotTask.isWorking: Boolean
    get() = activity == "working" || activity == "running" || busy == true

/** Waiting on a dispatched teammate (#1223). The live #1228 wire paints busy
 * and working during a coordination wait, so the flag outranks the painted
 * work: the row shows the wait, never the work spinner. */
val BotTask.isWaitingOnTeammate: Boolean
    get() = waitingOnTeammate == true

/** Running, needing the person, unread, or holding a queued send — client
 * state passed in, because the harness reports queues out-of-band, never as
 * task activity. */
fun BotTask.demandsAttention(queued: Boolean = false): Boolean =
    // The activity set is the BotActivity wire contract (working,
    // waiting-on-you, waiting, idle, no-signal, dead) plus the queued wait;
    // work states arrive through isWorking.
    isWaitingOnTeammate || isWorking || busy == true || unread == true || queued ||
        activity in setOf("waiting-on-you", "waiting", "queued")

/**
 * Attention outranks recency within a bot: waiting-on-you needs the person
 * most, then working/busy, then queued, then unread. A held send is client
 * state, so it ranks in the queued tier the way the wire value does. The
 * thread being looked at rides just above the idle tail; idle threads keep
 * stored order. Mirrors the desktop's orderedSidebarThreads so the tree, the
 * sheet, and the pickers agree on one order.
 */
fun attentionRank(task: BotTask, activeThreadId: String, queued: Boolean = false): Int = when {
    task.activity == "waiting-on-you" -> 0
    task.busy == true || task.activity == "working" -> 1
    task.activity == "queued" || queued -> 2
    task.unread == true -> 3
    task.threadId == activeThreadId -> 4
    else -> 5
}

/** Order, never filter: whatever the caller passes stays visible, only the
 * position changes. Sorting is stable, so equal ranks keep stored order. */
fun orderedThreads(
    tasks: List<BotTask>,
    activeThreadId: String,
    queuedThreadIds: Set<String> = emptySet(),
): List<BotTask> =
    tasks.sortedBy { attentionRank(it, activeThreadId, queued = it.threadId in queuedThreadIds) }

/** Routine results are ordinary threads; only their internal per-run executions are hidden. */
val Bot.visibleTasks: List<BotTask>
    get() = tasks.orEmpty().filter { it.routineRunId == null }

/**
 * Preserve saved folder order; attention floats threads within each group.
 * A missing folder leaves its threads unfiled. Search includes closed threads
 * and matches folder names, and keeps relevance (stored) order.
 */
fun Bot.threadGroups(
    matching: String = "",
    includingClosed: Boolean = false,
    /** Threads holding a queued send. A closed thread with a held send stays
     * in the list the way a running one does (Sidebar.tsx 865). */
    queuedThreadIds: Set<String> = emptySet(),
): List<BotThreadGroup> {
    val search = matching.trim()
    val threads = when {
        tasks == null -> listOf(BotTask(
            threadId = threadId, title = "", createdAt = createdAt,
            modelSelection = modelSelection, busy = busy, activity = activity, unread = unread,
            waitingOnTeammate = waitingOnTeammate,
            approvalMode = approvalMode, autoApprove = autoApprove, alwaysAllow = alwaysAllow,
        ))
        includingClosed || search.isNotEmpty() -> visibleTasks
        // Closed and archived threads fold away with the same override: one
        // that starts working, waits on the person, turns unread, or is
        // holding a queued send is back.
        else -> visibleTasks.filter {
            (!it.isClosed && !it.isArchived) ||
                it.demandsAttention(queued = queuedThreadIds.contains(it.threadId)) ||
                it.threadId == threadId
        }
    }
    val ordered = if (search.isEmpty()) orderedThreads(threads, threadId, queuedThreadIds) else threads
    val projectIds = mutableSetOf<String>()
    val groups = buildList {
        projects.orEmpty().forEach { project ->
            if (projectIds.add(project.id)) {
                val filed = ordered.filter { it.projectId == project.id }
                if (filed.isNotEmpty()) add(BotThreadGroup(project, filed))
            }
        }
        val unfiled = ordered.filter { it.projectId !in projectIds }
        if (unfiled.isNotEmpty()) add(BotThreadGroup(null, unfiled))
    }
    if (search.isEmpty()) return groups
    return groups.mapNotNull { group ->
        if (group.project?.name?.contains(search, ignoreCase = true) == true) group
        else group.tasks.filter { it.displayTitle.contains(search, ignoreCase = true) }
            .takeIf { it.isNotEmpty() }?.let { BotThreadGroup(group.project, it) }
    }
}
